enhancements/installer/openstack-ipi-failure-domains.md

```diff
@@ -0,0 +1,260 @@
+
+## Motivation
+
+### User Stories
+
+- As an administrator, I would like to stretch my OpenShift control plane across multiple Failure Domains (x3) so I increase the cluster resilience and performance.
+- As an administrator, I would like to deploy my OpenShift workers across multiple Failure Domains (>1).
+
+### Goals
+
+- The user will be able to define Failure Domains in the `install-config.yaml` file used by the OpenShift Installer.
+- The user will have the ability to deploy an OpenShift cluster across multiple availability zones which have their own networks and storage resources.
+
+### Non-Goals
+
+- The networks and subnets have to be pre-created in OpenStack Neutron.
+- Routing between the networks is managed by the network infrastructure.
+- Control-Plane VIP (e.g. API, ingress) route advertisement is not discussed in this enhancement and will
+  be solved by another proposal. If needed, more options will be added to the Failure Domain block to allow more advanced configurations; but this is out of scope for now.
+
+## Proposal
+
+The Installer allow users to provide a list of Failure Domains which contain information about networking and storage.
+The cluster will be deployed in these domains on day 1 via IPI.
+The Installer will validate the information given by the user and return an error if a resource doesn't exist or
+is incorrectly used.
+
+The cluster will be deployed in the different Failure Domains, following the Round Robin process. The first master will be deployed in the first domain, and once we have used all domains, we come back to the first one, etc.
+
+When using Failure Domains, the users will have to do it for both the masters and the workers.
+
+The infrastructure resources owned by the cluster continue to be clearly identifiable and distinct from other resources.
+
+Destroying a cluster must make sure that no resources are deleted that didn't belong to the cluster.  Leaking resources is preferred over any possibility of deleting non-cluster resources.
+
+### Workflow Description
+
+- The OpenStack administrators will deploy OpenStack with multiple Availability Zones, so Nova and Cinder can respectfully deploy servers and volumes in the Failure Domain.
+- The OpenStack administrators will create at least one Routed Provider Network, and then multiple subnets, where each zone has a least one subnet.
+- The OpenShift administrators will identify the Domain Failures: their availability zones, subnets, etc. They'll decide which ones will be used for the masters, and the ones for the workers (domains can be shared).
+- The OpenShift administrators will provide the right configuration in the `install-config.yaml` file and then deploy the cluster (detailed later in this document).
+- (Out of scope for this enhancement, see "Non-Goals") The OpenShift administrators will make sure that the network infrastructure has a networking route to reach the OCP Control Plane VIPs.
+
+### API Extensions
+
+Not applicable.
```

Comment:
I don't have much background for the machine-api operator on OpenStack. How does the machine-api operator get Failure Domain Spec for the compute nodes?

Reply:
MAO already supports AZs and subnets, so we should be good, if I understood the question correctly.
@mandre can confirm

Reply:
I provided an example MachineSet for clarity.